# GWAS report

Quarto templates to generate reports from summary stats of GWAS or rare variant analysis.

The report uses the [gwaslab package](https://github.com/Cloufield/gwaslab/) to generate plots. 

## Quick start

You need [Quarto](https://quarto.org/) and [Papermill](https://papermill.readthedocs.io/en/latest/) installed on your system to render the report. Then with the following command you can generate a `gwas_report_template.html` report from a summary stats file, a phenotype file and a tophits table:

```bash
quarto render gwas_report_template.qmd \
	-P sumstat_file:"mysumstats.tsv" \
	-P annotated_toploci_filename:"toploci.tsv" \
	-P phenotype_file:"phenotypes.tsv" \
	--to html
```

We provide a docker image pre-packed with all necessary dependencies at `edg1983/gwas-rmd-reports:v2.0`

## GWAS report

See the example report `example_gwas_report.html`

The minimum required input is a summary stats file in one of the format supported by the [gwaslab package](https://github.com/Cloufield/formatbook) and the corresponding phenotype file. Both are expected to be tab-separated.

Additionally, you can provide a tophits table that identify top SNPs with overlapping genes annotations. This is a tab-separated file with header and the following columns: 'CHROM','GENPOS','SNPID','LOG10P','CLOSEST_GENE_NAME'.

Finally, the report can take a toploci file describing loci clumped from the summary stats. This is expected to be in the format generated by [plink --clump](https://www.cog-genomics.org/plink/1.9/postproc#clump). An header line must be present and minimum required columns are: 'CHR', 'POS', 'SNP'.

### Parameters

- project: report title
- date: date
- version: version for the analysis
- sumstat_file: summary stats file
- phenotype: phenotype name
- covariates: covariates names
- phenotype_file: file containing the phenotype data
- manhattan_annotation_type: how to annotate manhattan plot. Can be "genes" when a tophits table is provided or "rsid" when no tophits table is provided.
- annotation_min_log10p: an additional line is added to the manhattan plot at this value
- annotated_tophits_filename: annotated top hits table
- annotated_toploci_filename: top loci table
- max_loci: number of top loci to plot
- regional_plot_window_kb: kb to expand around each locus in regional plots
- genome_build: genome build. Can be hg19 or hg38
- sumstat_format: format of the summary stats file (see above)

The following parameters can be used to load tables generated from the log files of regenie step1 and step2 and the processing of covariates and phenotypes input files. They can be omitted.

- regenie_step1_log
- regenie_step2_log
- phenotype_log
- covariate_log